### PR TITLE
[WIP] ArcFace loss as a triplet.

### PR DESCRIPTION
**Pull request by b0nce**

### CODE REVIEW

**Comment by AlekseySh** [comment]
```
        # def cosine_cdist(a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
        #     product = a @ b.transpose(0, 1)
        #     # print(product.max())
        #     return torch.acos(product - 1e-5)
        # positive_dist = cosine_cdist(anchor, positive)
        # negative_dist = cosine_cdist(anchor, negative)
       ```
       
       metric was 0.37
       
       ```
       postfix: metric_learning

seed: 42
gpus:
  - 0
  - 1

dataframe_name: df_fixed_train_fixed_val.csv
dataset_root: /nydl/data/DeepFashion_InShop/
logs_root: /nydl/logs/cur/ml/deepfashion/
logs_folder: ${now:%Y-%m-%d_%H-%M-%S}_${postfix}

im_size: 224
im_pad_ratio_train: 0.0
im_pad_ratio_val: 0.0
num_workers: 20
cache_size: 100000

sampler:
  name: SequentialBalanceSampler
  args:
    p: 2
    k: 8

bs_val: 96
max_epochs: 10000
valid_period: 1

metric_args:
  cmc_top_k: [1]
  map_top_k: [5]

augs_key: all

model:
  name: vit
  args:
    normalise_features: True
    use_multi_scale: False
    weights: pretrained_dino
    arch: vits8
    strict_load: True

criterion:
  name: TripletLossWithMiner
  args:
    need_logs: True
    margin: 0.1
    reduction: mean
    miner:
      name: TripletMinerWithMemory
      args:
        bank_size_in_batches: 20
        tri_expand_k: 5

optimizer:
  name: adam
  args:
    lr: 1e-5  # if you provide scheduler this parameter will be ignored

scheduler: null


# To use neptune you should also specify NEPTUNE_API_TOKEN in
# .env file or via `export NEPTUNE_API_TOKEN=...`
neptune_project: NEWYORKER/similarity-api


hydra_dir: ${logs_root}/${logs_folder}/

pod: pod1

tags:
  - ${postfix}
  - ${pod}
  - deepfashion

hydra:
  run:
    dir: ${hydra_dir}
  searchpath:
    - file://../../configs

```

**Comment by AlekseySh** [comment]
```
import torch

def norm(x):
    xn = torch.linalg.norm(x, 2, dim=1).detach()
    x = x.div(xn.unsqueeze(1))
    return x


d = 128
bs = 512
anchor = norm(torch.randn(bs, d))
positive = norm(torch.randn(bs, d))
negative = norm(torch.randn(bs, d))



def cosine_cdist(a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
    cos_dist = a @ b.transpose(0, 1)
    print(cos_dist.max())
    
    acos_dist = torch.acos(cos_dist)
    
    print("acos ", acos_dist)
    print(torch.isnan(acos_dist).any())

    return acos_dist

# positive_dist = cosine_cdist(torch.tensor([[0, 1, 0]]).float(), torch.tensor([[1, 0, 0]]).float())

positive_dist = cosine_cdist(anchor, positive)

```
